lib/ex_unit/lib/ex_unit/runner.ex

```diff
@@ -323,18 +440,31 @@ defmodule ExUnit.Runner do
   end
 
   defp exec_on_exit(test_or_case, pid, timeout) do
-    case ExUnit.OnExitHandler.run(pid, timeout) do
-      :ok ->
-        test_or_case
-
-      {kind, reason, stack} ->
-        state = test_or_case.state || failed(kind, reason, prune_stacktrace(stack))
-        %{test_or_case | state: state}
+    if ExUnit.OnExitHandler.registered?(pid) do
+      case ExUnit.OnExitHandler.run(pid, timeout) do
+        :ok ->
+          test_or_case
+
+        {kind, reason, stack} ->
+          state = test_or_case.state || failed(kind, reason, prune_stacktrace(stack))
+          %{test_or_case | state: state}
+      end
+    else
+      test_or_case
     end
   end
 
   ## Helpers
 
+  defp max_failures_reached?(failure_counter, max_failures)
```

Comment:
What do you think about pattern-match on `:infinity` to speed up the happy case?

```elixir
defp max_failures_reached?(_, :infinity), do: false

defp max_failures_reached?(manager, max_failures) when is_tuple(manager) do
  get_failure_counter(manager) >= max_failures
end

defp max_failures_reached?(failures, max_failures), do: failures >= max_failures
```

Reply:
I'm not too sure about removing the type checking guards since we are not validating max_failures when the functions is called directly.

Reply:
That's also part of another comment [I was currently writing](https://github.com/elixir-lang/elixir/pull/8174/files#r217171086). If applied, it would ensure that `max_failures` is either `:infinity` or a positive integer 🙂